src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/DistributedCacheTagHelperService.cs

```diff
@@ -1,44 +1,61 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Concurrent;
 using System.IO;
+using System.Linq;
 using System.Text;
 using System.Text.Encodings.Web;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Html;
 using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.TagHelpers.Internal;
 using Microsoft.AspNetCore.Razor.TagHelpers;
 using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Logging;
 
 namespace Microsoft.AspNetCore.Mvc.TagHelpers.Cache
 {
     /// <summary>
-    /// Implements <see cref="IDistributedCacheTagHelperService"/> and ensure
+    /// Implements <see cref="IDistributedCacheTagHelperService"/> and ensures
     /// multiple concurrent requests are gated.
+    /// The entries are stored like this:
+    /// <list type="bullet">
+    /// <item>
+    /// <description>Int32 representing the hashed cache key size.</description>
+    /// </item>
+    /// <item>
+    /// <description>The UTF8 encoded hashed cache key.</description>
+    /// </item>
+    /// <item>
+    /// <description>The UTF8 encoded cached content.</description>
+    /// </item>
+    /// </list>
     /// </summary>
     public class DistributedCacheTagHelperService : IDistributedCacheTagHelperService
     {
         private readonly IDistributedCacheTagHelperStorage _storage;
         private readonly IDistributedCacheTagHelperFormatter _formatter;
         private readonly HtmlEncoder _htmlEncoder;
-        private readonly ConcurrentDictionary<string, Task<IHtmlContent>> _workers;
+        private readonly ILogger _logger;
+        private readonly ConcurrentDictionary<CacheTagKey, Task<IHtmlContent>> _workers;
 
         public DistributedCacheTagHelperService(
             IDistributedCacheTagHelperStorage storage,
             IDistributedCacheTagHelperFormatter formatter,
-            HtmlEncoder HtmlEncoder 
-        )
+            HtmlEncoder HtmlEncoder,
+            ILoggerFactory loggerFactory)
         {
             _formatter = formatter;
             _storage = storage;
             _htmlEncoder = HtmlEncoder;
-
-            _workers = new ConcurrentDictionary<string, Task<IHtmlContent>>();
+            _logger = loggerFactory.CreateLogger<DistributedCacheTagHelperService>();
+            _workers = new ConcurrentDictionary<CacheTagKey, Task<IHtmlContent>>();
         }
 
         /// <inheritdoc />
-        public async Task<IHtmlContent> ProcessContentAsync(TagHelperOutput output, string key, DistributedCacheEntryOptions options)
+        public async Task<IHtmlContent> ProcessContentAsync(TagHelperOutput output, CacheTagKey key, DistributedCacheEntryOptions options)
```

Comment:
Isn't there a reasonable way to break this method into smaller ones? It's 120 lines long with several if-else, while and try cach loops nested and It makes really hard to understand what's going on.